### ISSUE TITLE
Geometry_oM: Update Mesh class to carry vertex colours

### ISSUE THREAD

**Issue by IsakNaslundBh** [BHoM/BHoM#594]
#### Description:

To enable visualisation of results on meshes, vertex colours needs to be added to the Mesh class.

Proposal to add new class called Vertex with Point and System.Drawing.Colour as properties and update the mesh class to carry a `List<Vertex>` instead of `List<Point>`.

Explicit casting from Point to Vertex should also be added for the Vertex class.

FYI @pawelbaran @enarhi @al-fisher @rwemay @FraserGreenroyd @adecler

**Comment by adecler** [comment]
If there is an intent to add Colour as a property of the Mesh points, I wonder if it wouldn't be beneficial to address the question of storing colour on any geometry. If mesh is under the spotlight now, I don't see why other geometry wouldn't eventually be too. So might as well consider the larger problem straight away 😄

**Comment by al-fisher** [comment]
@adecler to your point above:
I so agree and think any geometrical object should be able to hold colour information. Might this be simply through the means of formalising a constant ColourFragment of some sort?

To host more complex colour patterns beyond a simple constant shade - I think perhaps you need to be a mesh to render it? 
Thus I think there is a specific use case to build out the mesh beyond a simple constant colour fragment for meshes holding colour information on its vertices/points in order for the mesh to represent gradated colour maps etc.  

@IsakNaslundBh @adecler does the above make sense?

**Comment by rwemay** [comment]
If speaking about visual representation (colour) maybe should extend to a couple of other parameters - line weights/thicknesses, line 'types' - dashed etc. These maybe a separate/complex set of parameters/issues related more to a 'formal' UI, so unrelated, but again, I think worth considering that how geometry is displayed could be a wider/bigger topic/framework.

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd reason for pushing this through to the core oM/Engine itself is to be able to return already coloured geometries from methods that has no knowledge of the UI whatsoever. and without requiring any additional UI fiddling after that component. Basically just plug in what you want to visualise into the component and get the result straight away, not just the components that lets you display it using for example a "Custom Preview" component in GH. For example being able to colour spaces based on usage or similar etc, if this make sense?

Yeah I get that @IsakNaslundBh I was referring specifically to @rwemay idea of line weight, dashed line type, etc. :smile:

**Comment by rwemay** [comment]
@IsakNaslundBh , I still think we need to address the wider context/system. What's a colour attached to a mesh in Grasshopper, could be a longer list of requirements/visual conditions in another platform/use case. Transparency etc. Agree with @FraserGreenroyd  that that is more at the level of the UI, but needing to call common render settings for geometry. Maybe need to consider differentiating between 'what an object is' and 'how an object renders/visualises'. I guess the obvious solution would be fragments on geometry objects as mentioned above, if we didn't want to constantly write separate methods to 'render' an object versus get it's geometry. Maybe worth a call/adding to an agenda?

**Comment by kThorsager** [comment]
One issue I see with wrapping a general geometry with a colour (or material) is that after that the object will be "dead". As no methods currently accept a `ColouredGeometry` and I assume that there are no plans to make all methods compatible(?).

A thought is to let a `ColouredGeometry` carry a `DisplayFragment` that it passes over to the T.ex. `Bar` it's connected to. Then methods like `CentreLine()` could put them together again as a `ColouredGeometry`. 
Haven't been deep enough in the BHoM to see the feasibility of this.

Apart from that the easy solution would bee to add it in `IGeometry` or similar, but I have heard that that suggestion is frowned upon.

On the other hand I'm guessing that there's not really any way to retain the colour through a Grasshopper Component regardless (and you can't do everything in BHoM yet(?)) so it might not be relevant, but I don't think that we in a long term should restrict us to the grasshopper concept that colours only exist with geometry in Rhino



But I think that keeping the coloured mesh as a special case is a good idea and don't think how that is handled will carry over that much on the general case of displaying geometry, so perhaps a new issue for that?

But I imagine that it would be good to have different levels of specification of the display and a split in how we handle surfaces and lines.
So here are some thoughts on how to structure it, where each level has the information from the previous levels
### Surfaces
  #### level 1
Diffuse Colour
  #### level 2
  Glossy Colour
  (Roughness/Hardness)
(Transparancy, the `System.Drawing.Color` has alpha but depending on how we handle it it might be nice with a switch or similar)
#### level 3
  [Open Shading Language](https://en.wikipedia.org/wiki/Open_Shading_Language)
  or similar (needs a fact check if this even is a good one but from the talk of different visualisation software's  it might be nice at some point in the future)

### Lines/Points
#### level 1
  Colour
#### level 2
  Line weight
  Line Type (dashed, etc)
#### level 3 (2?)
  Patching (?, not completely sure on how to relate to this as it's kind of a surface)

**Comment by al-fisher** [comment]
@kThorsager - I think your thoughts above are good. 

Specifically needing to appreciate the difference of approach between points, curves, surfaces and meshes.

I like the idea of a "Display Properties". Also picking up on your and @rwemay point that we do not wnat to have to have two systems - `GetGeometry` and any method operating on `IGeometry` needs to still work as expected.

> But I think that keeping the coloured mesh as a special case is a good idea and don't think how that is handled will carry over that much on the general case of displaying geometry, so perhaps a new issue for that?

I would agree - and do suggest we continue with close out of the mesh - and in parallel create a new issue for planning the wider general "Display" of geometries.

**Comment by IsakNaslundBh** [comment]
> If I may throw a curve ball:
Why do we need to store colour on the mesh in the first place? Isn't colour calculated from one or some of the properties of the objects we need to colour? Storing colour directly feels wrong to me in term of fixing the choice of colour scale, colour range,...

The reason I have behind wanting to store it is really _only_ to be able to return it from a BHoM method in any toolkit without any knowledge of UI render methods etc. Want to be able to have methods returning a coloured mesh according to some gradient map, based on some calculation. The colour does not need to be based on the properties of the object itself, for example FEMesh and MeshForce. 

But if we can solve this in a clean way without returning the mesh with colours on it, I would be more than fine with that.

**Comment by al-fisher** [comment]
As discussed @IsakNaslundBh @rwemay : solution is to create a `RenderMesh` object in the `Graphics` namespace which will reference lists of `Vertex` objects which contain Point and system.colour properties.

This new mesh will therefore facilitate visualisation and rendering of any complex gradated coloured objects. We can convert from the parent objects (say NURBS, Curves or even Beam, Wall, Duct objects) into `RenderMeshes` for display in any UI. 

The higher order display properties such as Material Glossiness, line dashing, texture mapping can be handled separately and owned by the parent objects as Fragments as needed potentially. To then convert to the `RenderMesh` as needed.

The `Vertex` will be flexible enough to define colour in space for all use cases. Be it a Point Cloud, gradient across NURBS, along curves by uv and t parameters, 3D meshing, screen pixels etc.

Suggest we raise new issue for further expansion of Display properties - non-critical.
And @kThorsager are you happy to make modifications to your [draft PR](https://github.com/BHoM/BHoM/pull/596) in line with above to close out creation of these new objects?
